File: src/main/java/sample/controller/dialogs/rule/RuleDialog.java
package sample.controller.dialogs.rule;

import javafx.geometry.Pos;
import javafx.scene.control.*;
import javafx.scene.input.KeyCode;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Region;
import javafx.stage.Window;
import sample.controller.dialogs.rule.hrot.*;
import sample.controller.dialogs.rule.isotropic.*;
import sample.controller.dialogs.rule.misc.AlternatingDialog;
import sample.controller.dialogs.rule.misc.OneDimensionalDialog;
import sample.controller.dialogs.rule.misc.TurmitesDialog;
import sample.controller.dialogs.rule.ruleloader.RuleLoaderDialog;
import sample.model.Coordinate;
import sample.model.rules.RuleFamily;

import java.util.Arrays;
import java.util.logging.Level;
import java.util.logging.LogManager;
import java.util.logging.Logger;

public class RuleDialog extends Dialog {
    private RuleFamily chosenRuleFamily;

    // Controls
    private final GridPane grid;
    private final Button confirmRuleButton;
    private final TextField rulestringField;
    private final ComboBox ruleFamilyCombobox;

    // Array to store rule widgets
    private final RuleWidget[] ruleWidgets = new RuleWidget[]{new HROTDialog(), new HROTHistoryDialog(),
            new HROTSymbiosisDialog(), new HROTDeadlyEnemiesDialog(), new HROTGenerationsDialog(),
            new HROTExtendedGenerationsDialog(), new IntegerHROTDialog(), new DeficientHROTDialog(),
            new HROTRegeneratingGenerationsDialog(), new MultistateCyclicHROTDialog(), new INTDialog(),
            new INTHistoryDialog(), new INTEnergeticDialog(), new INTGenerationsDialog(), new DeficientINTDialog(),
            new OneDimensionalDialog(), new TurmitesDialog(), new AlternatingDialog(), new RuleLoaderDialog()};

    public RuleDialog(String promptText) {
        super();

        // Set Titles
        super.setTitle("Set Rule");
        super.setResizable(true);

        // Create controls
        grid = new GridPane();
        grid.setAlignment(Pos.CENTER);
        grid.setHgap(5);
        grid.setVgap(5);

        // HBox to contain combobox and button
        HBox hBox = new HBox();
        hBox.setSpacing(5);
        grid.add(hBox, 0, 0);

        // Combobox to Choose Rule Family
        ruleFamilyCombobox = new ComboBox();
        ruleFamilyCombobox.setOnAction((event) -> changeRuleFamily(
                ruleWidgets[ruleFamilyCombobox.getItems().indexOf(ruleFamilyCombobox.getValue())]));
        hBox.getChildren().add(ruleFamilyCombobox);

        // Textbox for rulestring
        rulestringField = new TextField();
        rulestringField.setPromptText(promptText);
        rulestringField.textProperty().addListener((observable, oldValue, newValue) ->
                rulestringFieldChanged());  // Listen for changes to the text and update the combobox

        rulestringField.setOnKeyPressed(event -> {  // Enter as a substitute for okay
            if (event.getCode().equals(KeyCode.ENTER)) {
                confirmRule();
            }
        });

        grid.add(rulestringField, 0, 2);

        // Button the confirm the rulestring
        confirmRuleButton = new Button("Confirm Rule");
        confirmRuleButton.setOnAction((event) -> confirmRule());
        grid.add(confirmRuleButton, 0, 3);

        // Get rule family names & configure the button correctly
        ruleFamilyCombobox.setValue(ruleWidgets[0].toString());
        for (RuleWidget widget : ruleWidgets) {
            ruleFamilyCombobox.getItems().add(widget.toString());
        }

        // Button to display information about the rulespace
        Button infoButton = new Button("?");
        infoButton.setOnAction((event) -> showRuleFamilyDescription());
        hBox.getChildren().add(infoButton);

        // Allows closing with close button
        Window window = super.getDialogPane().getScene().getWindow();
        window.setOnCloseRequest(event -> window.hide());

        // Add the rule widget to the grid
        grid.add(ruleWidgets[0], 0, 1);
        super.getDialogPane().setContent(grid);
    }

    public RuleDialog() {
        this("Enter Rulestring");
    }

    public void confirmRule() {
        // Get currently selected rule widget
        int index = ruleFamilyCombobox.getItems().indexOf(ruleFamilyCombobox.getValue());
        RuleWidget ruleWidget = ruleWidgets[index];

        // Update the rule in the rule widget
        try {
            ruleWidget.updateRule(rulestringField.getText());
        } catch (IllegalArgumentException exception) {
            LogManager.getLogManager().getLogger(Logger.GLOBAL_LOGGER_NAME).log(
                    Level.WARNING, exception.getMessage());

            Alert alert = new Alert(Alert.AlertType.ERROR);
            alert.setTitle("Error!");
            alert.setHeaderText("This rulestring is invalid!");
            alert.setContentText(exception.getMessage() + "\nIf you suspect this is a bug, please report it!");
            alert.showAndWait();
            return;
        } catch (Exception exception) {
            LogManager.getLogManager().getLogger(Logger.GLOBAL_LOGGER_NAME).log(
                    Level.WARNING, exception.getMessage());
            exception.printStackTrace();

            Alert alert = new Alert(Alert.AlertType.ERROR);
            alert.setTitle("Error!");
            alert.setHeaderText("An error occured!");
            alert.setContentText(exception.getMessage() + "\nIf you suspect this is a bug, please report it!");
            alert.showAndWait();
            return;
        }

        chosenRuleFamily = ruleWidget.getRuleFamily();
        if (chosenRuleFamily.getBoundedGrid() != null) {
            chosenRuleFamily.getBoundedGrid().setInitialCoordinate(new Coordinate(1800, 1800));
        }

        // Canonise the rulestring in the text field
        rulestringField.setText(chosenRuleFamily.getRulestring());

        // Close the dialog
        super.setResult(Boolean.TRUE);
        super.close();
    }

    public void rulestringFieldChanged() {
        boolean found = false;
        for (RuleWidget widget: ruleWidgets) {
            for (String regex: widget.getRuleFamily().getRegex()) {
                if (rulestringField.getText().split(":")[0].matches(regex)) {
                    found = true;
                    break;
                }
            }

            // Completely break out of the loop
            if (found) {
                changeRuleFamily(widget);  // Change to the widget who regex matches
                break;
            }
        }
    }

    public void changeRuleFamily(RuleWidget widget) {
        // Get currently selected rule widget
        int index = ruleFamilyCombobox.getItems().indexOf(ruleFamilyCombobox.getValue());
        RuleWidget ruleWidget = ruleWidgets[index];

        // Remove the current widget
        grid.getChildren().remove(ruleWidget);

        // Add the new one
        grid.add(widget, 0, 1);

        // Set the value of the combobox
        ruleFamilyCombobox.setValue(widget.getRuleFamily().getName());
    }

    public void showRuleFamilyDescription() {
        // Get currently selected rule widget
        int index = ruleFamilyCombobox.getItems().indexOf(ruleFamilyCombobox.getValue());
        RuleWidget ruleWidget = ruleWidgets[index];

        // Display description
        Alert alert = new Alert(Alert.AlertType.INFORMATION);
        alert.setTitle(ruleWidget.getRuleFamily().getName() + " Description");
        alert.setHeaderText("Description of " + ruleWidget.getRuleFamily().getName());
        alert.setContentText(ruleWidget.getRuleFamily().getDescription());
        alert.getDialogPane().setMinHeight(Region.USE_PREF_SIZE);  // Makes it scale to the text
        alert.showAndWait();
    }

    public RuleFamily getRule() {
        if (super.getResult() == Boolean.TRUE) {
            return chosenRuleFamily;  // Only return if user hits confirm rule
        }
        else {
            return null;
        }
    }

    public RuleWidget[] getRuleWidgets() {
        return ruleWidgets;
    }

    public void setRule(RuleFamily rule) {
        boolean found = false;
        for (RuleWidget widget: ruleWidgets) {
            for (String regex: widget.getRuleFamily().getRegex()) {
                if (rule.getRulestring().split(":")[0].matches(regex)) {
                    found = true;
                    break;
                }
            }

            // Completely break out of the loop
            if (found) {
                rulestringField.setText(rule.getRulestring());
                widget.setRuleFamily(rule);
                changeRuleFamily(widget);  // Change to the widget whose regex matches

                // Move the initial coordinate
                if (rule.getBoundedGrid() != null) {
                    rule.getBoundedGrid().setInitialCoordinate(new Coordinate(1800, 1800));
                }

                break;
            }
        }
    }
}

File: src/main/kotlin/application/controller/dialogs/rule/RuleDialog.kt
package application.controller.dialogs.rule

import application.controller.dialogs.rule.hrot.*
import application.controller.dialogs.rule.isotropic.*
import application.controller.dialogs.rule.misc.*
import application.controller.dialogs.rule.ruleloader.RuleLoaderDialog
import application.model.Coordinate
import application.model.rules.RuleFamily
import javafx.beans.value.ObservableValue
import javafx.event.EventHandler
import javafx.geometry.Pos
import javafx.scene.control.*
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.layout.GridPane
import javafx.scene.layout.HBox
import javafx.scene.layout.Region
import java.lang.Boolean
import java.util.logging.Level
import java.util.logging.LogManager
import java.util.logging.Logger
import kotlin.Any
import kotlin.Exception
import kotlin.IllegalArgumentException
import kotlin.String
import kotlin.arrayOf

class RuleDialog @JvmOverloads constructor(promptText: String? = "Enter Rulestring") : Dialog<Any?>() {
    private var chosenRuleFamily: RuleFamily? = null

    // Controls
    private val grid: GridPane
    private val confirmRuleButton: Button
    private val rulestringField: TextField
    private val ruleFamilyCombobox: ComboBox<String>

    // Array to store rule widgets
    private val ruleWidgets = arrayOf(
        HROTDialog(), HROTHistoryDialog(),
        HROTSymbiosisDialog(), HROTDeadlyEnemiesDialog(), HROTGenerationsDialog(),
        HROTExtendedGenerationsDialog(), IntegerHROTDialog(), DeficientHROTDialog(),
        HROTRegeneratingGenerationsDialog(), HROTBSFKLDialog(), MultistateCyclicHROTDialog(),
        INTDialog(), INTHistoryDialog(), INTEnergeticDialog(), INTGenerationsDialog(),
        DeficientINTDialog(), EuclideanDialog(), OneDimensionalDialog(), TurmitesDialog(),
        MargolusDialog(), AlternatingDialog(), RuleLoaderDialog()
    )

    // Only return if user hits confirm rule
    var rule: RuleFamily?
        get() = if (super.getResult() === Boolean.TRUE) {
            chosenRuleFamily // Only return if user hits confirm rule
        } else null
        set(rule) {
            var found = false
            for (widget in ruleWidgets) {
                for (regex in widget.ruleFamily!!.regex) {
                    if (rule!!.rulestring.split(":").toTypedArray()[0].matches(Regex(regex))) {
                        found = true
                        break
                    }
                }

                // Completely break out of the loop
                if (found) {
                    rulestringField.text = rule!!.rulestring
                    widget.ruleFamily = rule
                    changeRuleFamily(widget) // Change to the widget whose regex matches

                    // Move the initial coordinate
                    if (rule.boundedGrid != null) {
                        rule.boundedGrid.initialCoordinate = Coordinate(1800, 1800)
                    }

                    break
                }
            }
        }

    init {
        // Set Titles
        super.setTitle("Set Rule")
        super.setResizable(true)

        // Create controls
        grid = GridPane()
        grid.alignment = Pos.CENTER
        grid.hgap = 5.0
        grid.vgap = 5.0

        // HBox to contain combobox and button
        val hBox = HBox()
        hBox.spacing = 5.0
        grid.add(hBox, 0, 0)

        // Combobox to Choose Rule Family
        ruleFamilyCombobox = ComboBox<String>()
        ruleFamilyCombobox.onAction = EventHandler {
            changeRuleFamily(
                ruleWidgets[ruleFamilyCombobox.items.indexOf(ruleFamilyCombobox.value)]
            )
        }
        hBox.children.add(ruleFamilyCombobox)

        // Textbox for rulestring
        rulestringField = TextField()
        rulestringField.promptText = promptText

        // Listen for changes to the text and update the combobox
        rulestringField.textProperty().addListener { _: ObservableValue<out String>, _: String, _: String -> rulestringFieldChanged() }
        rulestringField.onKeyPressed = EventHandler { event: KeyEvent ->   // Enter as a substitute for okay
            if (event.code == KeyCode.ENTER) {
                confirmRule()
            }
        }
        grid.add(rulestringField, 0, 2)

        // Button the confirm the rulestring
        confirmRuleButton = Button("Confirm Rule")
        confirmRuleButton.onAction = EventHandler { confirmRule() }
        grid.add(confirmRuleButton, 0, 3)

        // Get rule family names & configure the button correctly
        ruleFamilyCombobox.setValue(ruleWidgets[0].toString())
        for (widget in ruleWidgets) {
            ruleFamilyCombobox.items.add(widget.toString())
        }

        // Button to display information about the rulespace
        val infoButton = Button("?")
        infoButton.onAction = EventHandler { showRuleFamilyDescription() }
        hBox.children.add(infoButton)

        // Allows closing with close button
        val window = super.getDialogPane().scene.window
        window.onCloseRequest = EventHandler { window.hide() }

        // Add the rule widget to the grid
        grid.add(ruleWidgets[0], 0, 1)
        super.getDialogPane().content = grid
    }

    private fun confirmRule() {
        // Get currently selected rule widget
        val index = ruleFamilyCombobox.items.indexOf(ruleFamilyCombobox.value)
        val ruleWidget = ruleWidgets[index]

        // Update the rule in the rule widget
        try {
            ruleWidget.updateRule(rulestringField.text)
        } catch (exception: IllegalArgumentException) {
            LogManager.getLogManager().getLogger(Logger.GLOBAL_LOGGER_NAME).log(
                Level.WARNING, exception.message
            )

            val alert = Alert(Alert.AlertType.ERROR)
            alert.title = "Error!"
            alert.headerText = "This rulestring is invalid!"
            alert.contentText = """
                ${exception.message}
                If you suspect this is a bug, please report it!
                """.trimIndent()
            alert.showAndWait()
            return
        } catch (exception: Exception) {
            LogManager.getLogManager().getLogger(Logger.GLOBAL_LOGGER_NAME).log(
                Level.WARNING, exception.message
            )
            exception.printStackTrace()

            val alert = Alert(Alert.AlertType.ERROR)
            alert.title = "Error!"
            alert.headerText = "An error occured!"
            alert.contentText = """
                ${exception.message}
                If you suspect this is a bug, please report it!
                """.trimIndent()
            alert.showAndWait()
            return
        }

        chosenRuleFamily = ruleWidget.ruleFamily
        if (chosenRuleFamily!!.boundedGrid != null) {
            chosenRuleFamily!!.boundedGrid.initialCoordinate = Coordinate(1800, 1800)
        }

        // Canonise the rulestring in the text field
        rulestringField.text = chosenRuleFamily!!.rulestring

        // Close the dialog
        super.setResult(Boolean.TRUE)
        super.close()
    }

    private fun rulestringFieldChanged() {
        var found = false
        for (widget in ruleWidgets) {
            for (regex in widget.ruleFamily!!.regex) {
                if (rulestringField.text.split(":").toTypedArray()[0].matches(Regex(regex))) {
                    found = true
                    break
                }
            }

            // Completely break out of the loop
            if (found) {
                changeRuleFamily(widget) // Change to the widget who regex matches
                break
            }
        }
    }

    private fun changeRuleFamily(widget: RuleWidget) {
        // Get currently selected rule widget
        val index = ruleFamilyCombobox.items.indexOf(ruleFamilyCombobox.value)
        val ruleWidget = ruleWidgets[index]

        // Remove the current widget
        grid.children.remove(ruleWidget)

        // Add the new one
        grid.add(widget, 0, 1)

        // Set the value of the combobox
        ruleFamilyCombobox.value = widget.ruleFamily!!.name
    }

    private fun showRuleFamilyDescription() {
        // Get currently selected rule widget
        val index = ruleFamilyCombobox.items.indexOf(ruleFamilyCombobox.value)
        val ruleWidget = ruleWidgets[index]

        // Display description
        val alert = Alert(Alert.AlertType.INFORMATION)
        alert.title = ruleWidget.ruleFamily!!.name + " Description"
        alert.headerText = "Description of " + ruleWidget.ruleFamily!!.name
        alert.contentText = ruleWidget.ruleFamily!!.description
        alert.dialogPane.minHeight = Region.USE_PREF_SIZE // Makes it scale to the text
        alert.showAndWait()
    }
}